### ISSUE TITLE
support Lemmy

### ISSUE THREAD

**Issue by snarfed** [snarfed/bridgy-fed#372]
Background:

- https://join-lemmy.org/
- https://join-lemmy.org/docs/en/administration/federation_getting_started.html
- https://socialhub.activitypub.rocks/t/help-improving-federation-between-lemmy-and-other-projects/2308

I tried federating [this like](https://snarfed.org/2023-01-17_lemmy-ml) to [this post](https://lemmy.ml/post/713841) on lemmy.ml (running BE 0.16.7) just now, and the AP inbox delivery to `https://lemmy.ml/u/OptimusPrime/inbox` returned 400 with `invalid type: map, expected a string representing an URL at line 1 column 218`. [Bridgy Fed log here.](https://fed.brid.gy/log?start_time=1673989596&key=https://snarfed.org/2023-01-17_lemmy-ml%20https://lemmy.ml/post/713841&module=) Full AS2 object below.

Lemmy's design looks like Reddit, with up/downvotes instead of likes, so I don't know that it actually supports likes, but the error message sounds unrelated to that, more like maybe it can only handle string URLs for `actor`s instead of compound objects.

```json
{
  "published": "2023-01-17T13:03:52-08:00",
  "content": "likes <a class=\"u-like u-like-of\" href=\"https://lemmy.ml/post/713841\">[lemmy.ml](http://lemmy.ml/)</a>",
  "url": "https://fed.brid.gy/r/https://snarfed.org/2023-01-17_lemmy-ml",
  "actor": {
    "url": "https://fed.brid.gy/r/https://snarfed.org/",
    "image": {
      "url": "https://secure.gravatar.com/avatar/947b5f3f323da0ef785b6f02d9c265d6?s=96&d=blank&r=g",
      "type": "Image"
    },
    "type": "Person",
    "name": "Ryan Barrett",
    "icon": {
      "url": "https://secure.gravatar.com/avatar/947b5f3f323da0ef785b6f02d9c265d6?s=96&d=blank&r=g",
      "type": "Image"
    },
    "id": "https://fed.brid.gy/snarfed.org",
    "preferredUsername": "[snarfed.org](http://snarfed.org/)"
  },
  "@context": "https://www.w3.org/ns/activitystreams",
  "type": "Like",
  "object": "https://lemmy.ml/post/713841",
  "id": "https://fed.brid.gy/r/https://snarfed.org/2023-01-17_lemmy-ml",
  "cc": [
    "https://lemmy.ml/u/OptimusPrime",
    "https://lemmy.ml/c/privacy",
    "https://www.w3.org/ns/activitystreams#Public"
  ],
  "to": [
    "https://www.w3.org/ns/activitystreams#Public"
  ]
}
```

**Comment by snarfed** [comment]
From https://socialhub.activitypub.rocks/t/help-improving-federation-between-lemmy-and-other-projects/2308/22 :

> You are right this isnt complicated, in fact we are already doing the same thing in some other places. Unfortunately I dont have time to work on this, but if you want to make a contribution I would gladly guide you through the code.

**Comment by snarfed** [comment]
Progress is a new error message. After #442, we're now getting HTTP 400 `data did not match any variant of untagged enum PersonOrGroup` on this `Like`:

```json
{
  "published": "2023-10-11T21:12:28-07:00",
  "content": "Likes <a class=\"u-like-of\" href=\"https://lemmy.world/post/6663494\">https://lemmy.world/post/6663494</a>;..\n<div class=\"e-content\">\nA meme for math people – Lemmy.world\n</div>",
  "url": "https://fed.brid.gy/r/https://snarfed.org/2023-10-11_a-meme-for-math-people-lemmy-world",
  "actor": "https://fed.brid.gy/snarfed.org",
  "id": "https://fed.brid.gy/r/https://snarfed.org/2023-10-11_a-meme-for-math-people-lemmy-world",
  "@context": "https://www.w3.org/ns/activitystreams",
  "type": "Like",
  "object": "https://lemmy.world/post/6663494",
  "cc": [
    "https://lemmy.ml/c/memes",
    "https://lemmy.world/u/hypertown",
    "https://www.w3.org/ns/activitystreams#Public"
  ],
  "to": [
    "https://www.w3.org/ns/activitystreams#Public"
  ]
}
```

**Comment by snarfed** [comment]
And for this reply, `400 Client Error: Bad Request for url: https://lemmy.world/u/hypertown/inbox ; deserializing body: ...`

```json
{
  "id": "https://fed.brid.gy/r/https://snarfed.org/2023-10-11_a-meme-for-math-people-lemmy-world-3#bridgy-fed-create",
  "actor": "https://fed.brid.gy/snarfed.org",
  "published": "2023-10-12T04:17:52.463875+00:00",
  "@context": "https://www.w3.org/ns/activitystreams",
  "type": "Create",
  "object": {
    "published": "2023-10-11T21:17:47-07:00",
    "content": "<a class=\"u-in-reply-to\" href=\"https://lemmy.world/post/6663494\"></a>\n<div class=\"e-content\">\nnice\n</div>",
    "url": "https://fed.brid.gy/r/https://snarfed.org/2023-10-11_a-meme-for-math-people-lemmy-world-3",
    "id": "https://fed.brid.gy/r/https://snarfed.org/2023-10-11_a-meme-for-math-people-lemmy-world-3",
    "type": "Note",
    "attributedTo": "https://fed.brid.gy/snarfed.org",
    "inReplyTo": "https://lemmy.world/post/6663494",
    "cc": [
      "https://lemmy.ml/c/memes",
      "https://lemmy.world/u/hypertown",
      "https://www.w3.org/ns/activitystreams#Public"
    ],
    "tag": [
      {
        "type": "Mention",
        "href": "https://lemmy.world/u/hypertown"
      }
    ],
    "to": [
      "https://www.w3.org/ns/activitystreams#Public"
    ]
  },
  "to": [
    "https://www.w3.org/ns/activitystreams#Public"
  ]
}
```

**Comment by snarfed** [comment]
> Progress is a new error message. After #442, we're now getting HTTP 400 `data did not match any variant of untagged enum PersonOrGroup` on this `Like`:

This also happens on follows.

**Comment by snarfed** [comment]
More on this error:
* https://socialhub.activitypub.rocks/t/we-have-created-documentation-on-how-exactly-lemmy-federation-works/1085/46?u=snarfed
* https://github.com/LemmyNet/Lemmy/issues/2756

**Comment by gaviscapes** [comment]
Is there any progress on this at all? The recent lemmy update made some heavy changes and improved the API significantly.

**Comment by snarfed** [comment]
Exciting! Last I checked, we were blocked on Lemmy errors, details above. Feel free to test again and see if anything has changed!

**Comment by Nutomic** [comment]
Lemmy maintainer here, just came across this. I threw [this activity](https://github.com/snarfed/bridgy-fed/issues/372#issuecomment-1758889702) into Lemmy tests, and it fails because the `Create` activity is missing `cc` field. You need to copy that from the inner `Note`.

By the way, Lemmy users view Bridgy Fed generally positively, as they consider all posts to be public anyway. So theres no need for complicated opt-in solutions in case of Lemmy. [Here you can see the discussion](https://lemmy.ml/post/11857402).

**Comment by Tamschi** [comment]
@snarfed This sound's actionable (or possibly even solved? I think something similar came up regarding another software), so I'm removing the `blocked` label since we should at least re-test if the same snag is being hit.

**Comment by snarfed** [comment]
Yes! We now duplicate to and cc between object and activity, so this might now work. Worth re-testing!

**Comment by mapto** [comment]
@Nutomic, could you repeat the test, please? This seems to be hanging in the air.

And it might be the leap towards resolving #1435 too.

**Comment by Nutomic** [comment]
@mapto Show me the json data and I will throw it into the tests.

**Comment by snarfed** [comment]
Thanks for the nudge, guys!

I tried delivering a `Like` and a `Follow` to lemmy.ml again just now, and it 400ed both with the error `{"error":"unknown","message":"Failed to parse object https://fed.brid.gy/snarfed.org  with content ..."}`. Not sure why it's unhappy with that actor, most other fediverse software is ok with it. Here's the full object, https://browser.pub/https://fed.brid.gy/snarfed.org :

```json
{
  "@context": [
    "https://www.w3.org/ns/activitystreams",
    "https://purl.archive.org/miscellany",
    {
      "discoverable": "http://joinmastodon.org/ns#discoverable",
      "indexable": "http://joinmastodon.org/ns#indexable"
    },
    {
      "PropertyValue": "http://schema.org#PropertyValue"
    },
    "https://w3id.org/security/v1",
    {
      "alsoKnownAs": {
        "@id": "as:alsoKnownAs",
        "@type": "@id"
      }
    }
  ],
  "type": "Person",
  "id": "https://fed.brid.gy/snarfed.org",
  "url": "https://fed.brid.gy/r/https://snarfed.org/about",
  "name": "Ryan Barrett",
  "alsoKnownAs": ["https://snarfed.org/"],
  "attachment": [{
      "name": "Ryan Barrett",
      "type": "PropertyValue",
      "value": "<a rel=\"me\" href=\"https://snarfed.org/about\"><span class=\"invisible\">https://</span>snarfed.org/about</a>"
    }],
  "discoverable": true,
  "endpoints": {
    "sharedInbox": "https://fed.brid.gy/ap/sharedInbox"
  },
  "followers": "https://fed.brid.gy/snarfed.org/followers",
  "following": "https://fed.brid.gy/snarfed.org/following",
  "icon": {
    "name": "Ryan",
    "type": "Image",
    "url": "https://snarfed.org/ryan_profile_medium.jpg"
  },
  "image": [
    {
      "type": "Image",
      "url": "https://snarfed.org/w/wp-content/uploads/2013/06/copy-cropped-iraq_bar.jpg"
    },
    {
      "name": "Ryan",
      "type": "Image",
      "url": "https://snarfed.org/ryan_profile_medium.jpg"
    }
  ],
  "inbox": "https://fed.brid.gy/snarfed.org/inbox",
  "indexable": true,
  "manuallyApprovesFollowers": false,
  "outbox": "https://fed.brid.gy/snarfed.org/outbox",
  "preferredUsername": "snarfed.org",
  "publicKey": {
    "id": "https://fed.brid.gy/snarfed.org#key",
    "owner": "https://fed.brid.gy/snarfed.org",
    "publicKeyPem": "-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAxi54qQrNQOcOnUftXBQm\n9BJoSCTOF6ffbVL+zXkEgktubVy5XD1Hli07CgvpDBz9jlKJNp1BJ2G0lpn0QBtW\nYvViPwlsgB1tsheChL3KtWTrnNl01+2lbYD2kgdTDXc/j/1BkL7ootHulE8jdjfK\ndNUUFoTmzF/5wgueNYMWNnWVSzGUVGyDo6TPYV6tmso3QRi0Cr5Sk28VqJ49oT5K\nbCv4eMrGJifbmVK4fhuTndQWeq1919pxnUByP5rsOBoj9b3YshygiWQAjaqV8W7W\n+kfybE4dGx4Cyj02HcgPlhgGexhSWo6Fpe9IvZGrKquql+kLyWRXvW5eQhb+l4bx\nlwIDAQAB\n-----END PUBLIC KEY-----"
  },
  "summary": "",
  "updated": "2025-09-05T19:23:00.070309+00:00"
}
```

**Comment by Nutomic** [comment]
I see now, the problem is that the current stable version of Lemmy doesnt support `image` having multiple items. This is already fixed in the upcoming version 1.0 but that will still take a long time to release. For now Im adding a change for the current release version to ignore images which cannot be parsed (https://github.com/LemmyNet/lemmy/pull/6015), though it may still take a month or so until we make a new release. Until then you can test with a perosn that doesnt send any `image`. Or use [voyager.lemmy.ml](https://voyager.lemmy.ml/u/snarfed.org@fed.brid.gy) which runs the latest dev version.

**Comment by mapto** [comment]
This issue seems to be resolved on the Lemmy side. Could we retest?

I'd like to follow a lemmy community from bluesky? Let's see how far we could get.

**Comment by snarfed** [comment]
@mapto please do! I'm not very familiar with Lemmy or the fediverse or ActivityPub groups yet. Feel free to try!

**Comment by mapto** [comment]
I would but I don't understand how. Could you explain the user experience?

Do I need to federate the group/community by following @bsky.brid.gy to replicate the mastodon procedure? Normally Communities/groups on Lemmy can't follow users.

Or is the way to find the Lemmy group from bluesky different?

In my head it would be reasonable that groups federate automatically as long as they are public, as was also for users in the early days of BridgyFed. Almost like hashtags.

**Comment by snarfed** [comment]
We haven't really figured out the UX for bridging fediverse groups yet; I guess that's part of what this issue tracks. If you're an admin for a group that you want to try bridging, let me know and I can try to enable it manually!

**Comment by snarfed** [comment]
Otherwise, I re-tested basic interaction just now by liking and commenting on https://lemmy.dbzer0.com/post/61403610, and those both delivered ok and show up on that post! 🎉 @Nutomic thanks again for your fix(es) here a while ago.

**Comment by flamingos-cant** [comment]
@snarfed I'd be happy to let you test this for [!unitedkingdom](https://feddit.uk/c/unitedkingdom) or [!uk_politics](https://feddit.uk/c/uk_politics).

**Comment by rimu** [comment]
If you'd like to try with PieFed, please use https://crust.piefed.social/c/test

**Comment by Die4Ever** [comment]
> Otherwise, I re-tested basic interaction just now by liking and commenting on https://lemmy.dbzer0.com/post/61403610, and those both delivered ok and show up on that post! 🎉 [@Nutomic](https://github.com/Nutomic) thanks again for your fix(es) here a while ago.

The comment seems to not federate across Lemmy properly, searching for the comment URL fails to pull it 

https://fed.brid.gy/r/https://snarfed.org/2026-01-11_which-toothpaste-brand-did-you-put-in-your-coffee-today-lemmy-2

**Comment by Die4Ever** [comment]
The test comment doesn't show on different Lemmy instances 

https://retrolemmy.com/post/31790836

Nor on PieFed https://piefed.social/c/shittyasklemmy/p/1650638/which-toothpaste-brand-did-you-put-in-your-coffee-today

**Comment by snarfed** [comment]
@Die4Ever is this the fediverse's usual missing replies problem? Bridgy Fed delivered the comment to its author's `sharedInbox`, https://lemmy.dbzer0.com/inbox. I think I've gotten [AS2 audience targeting](https://www.w3.org/TR/activitystreams-vocabulary/#audienceTargeting) mostly right for the object types Bridgy Fed handles so far, but I don't handle the [`audience` property](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-audience) specifically, which I may need to for groups posts...?Ugh, AS2 audience targeting is painful, but I can dive back in eventually if I need to.

**Comment by snarfed** [comment]
@flamingos-cant @rimu thanks for the offers! Can you all confirm that you own those admin accounts somehow? Maybe DM me at @snarfed@indieweb.social from your Lemmy accounts, or from the Mastodon accounts that you've linked to in your Lemmy profiles?

**Comment by Die4Ever** [comment]
> [@Die4Ever](https://github.com/Die4Ever) is this the fediverse's usual missing replies problem? Bridgy Fed delivered the comment to its author's `sharedInbox`, https://lemmy.dbzer0.com/inbox. I think I've gotten [AS2 audience targeting](https://www.w3.org/TR/activitystreams-vocabulary/#audienceTargeting) mostly right for the object types Bridgy Fed handles so far, but I don't handle the [`audience` property](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-audience) specifically, which I may need to for groups posts...?Ugh, AS2 audience targeting is painful, but I can dive back in eventually if I need to.

it's not a normal issue, usually searching for the APubID URL (`https://fed.brid.gy/r/https://snarfed.org/2026-01-11_which-toothpaste-brand-did-you-put-in-your-coffee-today-lemmy-2`) would retrieve the comment remotely even if it wasn't already pulled in via subscription

But both instances have local subscribers anyways, which explains why all the other comments are there except yours

**Comment by flamingos-cant** [comment]
> @Die4Ever is this the fediverse's usual missing replies problem? Bridgy Fed delivered the comment to its author's sharedInbox, https://lemmy.dbzer0.com/inbox. I think I've gotten [AS2 audience targeting](https://www.w3.org/TR/activitystreams-vocabulary/#audienceTargeting) mostly right for the object types Bridgy Fed handles so far, but I don't handle the [audience property](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-audience) specifically, which I may need to for groups posts...?Ugh, AS2 audience targeting is painful, but I can dive back in eventually if I need to.

This issue is that neither the `cc` or `to` fields contain the group ID.

Edit: Actually Lemmy gets the community from the post, the comment actually shows up for me: https://feddit.uk/post/42478096/22493582

Edit 2:  retrolemmy is on an old version of Lemmy that doesn't have Nutonic's fix deployed.

**Comment by snarfed** [comment]
@flamingos-cant thanks! Do you know if the key standard here is https://codeberg.org/fediverse/fep/src/branch/main/fep/1b12/fep-1b12.md ? or is it some other part of AS2 or AP that I'm not familiar with?

Looking at https://codeberg.org/fediverse/fep/src/branch/main/fep/1b12/fep-1b12.md#threads-and-comments , the only MUST it has for comments/replies is `inReplyTo`. It mentions `audience`, but doesn't say it's required, and doesn't mention `to`/`cc` at all.

Do groups forward comments to all followers with `Announce`s just like they forward top-level posts? https://codeberg.org/fediverse/fep/src/branch/main/fep/1b12/fep-1b12.md#the-announce-activity

**Comment by snarfed** [comment]
ok @flamingos-cant @rimu! I've bridged https://bsky.app/profile/unitedkingdom.feddit.uk.ap.brid.gy and https://bsky.app/profile/did:plc:dj2fncinw4wbtdavxxfz6c56 (Bluesky is behind on its handle, sorry, should resolve eventually). Most posts probably won't make it through, since AP groups posts are authored by their original authors, not by the group, and Bridgy Fed only bridges posts from bridged accounts. Still, let's see how it goes!

**Comment by flamingos-cant** [comment]
> @flamingos-cant thanks! Do you know if the key standard here is https://codeberg.org/fediverse/fep/src/branch/main/fep/1b12/fep-1b12.md ? or is it some other part of AS2 or AP that I'm not familiar with?
>
> Looking at https://codeberg.org/fediverse/fep/src/branch/main/fep/1b12/fep-1b12.md#threads-and-comments , the only MUST it has for comments/replies is inReplyTo. It mentions audience, but doesn't say it's required, and doesn't mention to/cc at all.

No your right, replies don't require the `audience` property or to `cc` the community. It was just the first thing that jumped out to me as missing because they are required for top level posts. The issue here is just old versions of Lemmy. Any instance on version < `0.19.14` [lacks the comment](https://sh.itjust.works/post/53210388), while up to date instances [have the comment](https://discuss.online/post/33670137). I don't know why it isn't working with Piefed, the devs would have to look at that.

> Do groups forward comments to all followers with Announces just like they forward top-level posts? https://codeberg.org/fediverse/fep/src/branch/main/fep/1b12/fep-1b12.md#the-announce-activity

Groups forward all activities (that the instance doesn't reject or errors out serialising) to its followers through `Announce`s. This is how Lemmy/Piefed/Mbin are able to keep things like votes and comments in sync between instances.

**Comment by Die4Ever** [comment]
> The test comment doesn't show on different Lemmy instances
> 
> https://retrolemmy.com/post/31790836
> 
> Nor on PieFed https://piefed.social/c/shittyasklemmy/p/1650638/which-toothpaste-brand-did-you-put-in-your-coffee-today

Manually pulling the comment works after updating to Lemmy 0.19.5

IDK about PieFed, as PieFed does not seem to have a way to manually pull a remote comment (only posts)

**Comment by flamingos-cant** [comment]
Posting to a Lemmy community from Blueksy works! https://feddit.uk/post/42603656 Formatting seems to have gone wrong somewhere though.

@snarfed Would it be possible to move the `Article` in the `tag` field to the `attachment` field? That's where Lemmy reads links from, so putting it there would have a post appear more Lemmy-native. Mastodon is also looking at using `Link`s in the `attachment` field as a way to generate preview cards for posts (mastodon/Mastodon#36104).

**Comment by Nutomic** [comment]
Question, how does the opt-in or opt-out to Bridgyfed work for Lemmy users? The [docs](https://fed.brid.gy/docs#fediverse-get-started) mention login and enable (which does not support Lemmy), or to follow `@bsky.brid.gy@bsky.brid.gy` (Lemmy does not support following users). In general I believe that Lemmy users understand their posts to be completely public, and are fine with having them bridged to Bluesky by default (contrary to Mastodon users, https://github.com/snarfed/bridgy-fed/issues/835). So I believe it is fine to enable bridging by default for all Lemmy instances and users, with opt-out done via instance block from Lemmy admins, or via fed.brid.gy website.

On another note, Bridgefed strips all formatting from posts which are bridged from [Bluesky](https://bsky.app/profile/flamingos-cant.xyz/post/3mcd423kj2c2t) to [Lemmy](https://feddit.uk/post/42603656). This is a problem because Lemmy takes only the first line of a microblogging post as title. As Bridgefed crams the entire content into a single `<p>` tag, it results in a weirdly formatted title (see Lemmy link above).

**Comment by snarfed** [comment]
> [@snarfed](https://github.com/snarfed) Would it be possible to move the `Article` in the `tag` field to the `attachment` field?

Yes! Good point, we can definitely look into that. Thanks for the nudge!

More broadly though, microblogging is obviously pretty different from groups/forums, I expect there's a limit on how well we'll be able to bridge from/to Bluesky specifically. We haven't done a deep dive into ATProto group/forum apps, but those might be more appropriate for bridging, #1178 notwithstanding. Feel free to look and let us know what you find!

cc @anujahooja

**Comment by snarfed** [comment]
@Nutomic thanks for looking at this! We'd love to get bridging working well for Lemmy and other fediverse groups servers. I appreciate your help here!

> Question, how does the opt-in or opt-out to Bridgyfed work for Lemmy users?

Good question! Can Lemmy users DM? If they DM `start` to @bsky.brid.gy@bsky.brid.gy, that will bridge them too. Regardless, you're right, it sounds like it's either not easy or not possible right now. 😕 

I don't know that we're ready to automatically enable the bridge for all Lemmy users, but we're definitely open to instance admins enabling it for their users! We've already worked with a number of instances to do that. Details in https://fed.brid.gy/docs#enable-api . The mechanism right now is still following, so we'd need to come up with a new one that works for Lemmy, but that's doable. If you know any instance admins who are interested, please send them our way, we're happy to talk! cc @anujahooja

>On another note, Bridgefed strips all formatting from posts which are bridged from [Bluesky](https://bsky.app/profile/flamingos-cant.xyz/post/3mcd423kj2c2t) to [Lemmy](https://feddit.uk/post/42603656)

We definitely don't strip formatting; we try hard to preserve it as faithfully as possible. Looking at https://browser.pub/https://feddit.uk/post/42603656 , the HTML content does have a single `<p>` tag, as you mention, but it includes the line break, links, and @-mention from the [original post on Bluesky](https://bsky.app/profile/did:plc:vtirffkcn4qsaendh6hb4lql/post/3mcd423kj2c2t). 

https://join-lemmy.org/docs/contributors/05-federation.html is really helpful for understanding Lemmy's AP profile and how it does federation. Thank you for that! I don't see anything about this first-line logic there, though, or in FEP-1b12. Is it documented anywhere else? Do other fediverse groups projects like Mbin/Kbin do it too?

**Comment by Die4Ever** [comment]
I think the opt-in should be by community, not by user. But I guess that might be hard cause it's so different from Mastodon.

We can DM, but it seems like there won't be much content if it requires both the user and the community to opt-in manually. I just sent the start DM. IDK if Lemmy DMs are the same as Mastodon DMs, but I sent it from `@Die4Ever@retrolemmy.com`

No DM response yet.

**Comment by Nutomic** [comment]
Lemmy currently doesnt use the standard `Note` type for DMs, but sends them as [ChatMessage](https://join-lemmy.org/docs/contributors/05-federation.html#private-message). This will be changed to `Note` in the upcoming 1.0 release (which is still a few months out).

The approach mentioned in the docs with `HTTP POST to https://fed.brid.gy/web-site` may also work (Lemmy sends out webmentions for new posts), though I dont see how you validate that the request actually comes from an instance admin. And it would be helpful if you include an example `curl` command for this in the docs.

> We definitely don't strip formatting; we try hard to preserve it as faithfully as possible. Looking at https://browser.pub/https://feddit.uk/post/42603656 , the HTML content does have a single <p> tag, as you mention, but it includes the line break, links, and @-mention from the [original post on Bluesky](https://bsky.app/profile/did:plc:vtirffkcn4qsaendh6hb4lql/post/3mcd423kj2c2t).

The `content` does include the line breaks as `\n`, but because this is HTML newlines are ignored and converted to a single space. Below it also uses `<br>`, these show up as newlines in Lemmy as expected. So you need to replace the remaining `\n` with `<br>` as well (or use a separate `<p>` tag for each paragraph).

> https://join-lemmy.org/docs/contributors/05-federation.html is really helpful for understanding Lemmy's AP profile and how it does federation. Thank you for that! I don't see anything about this first-line logic there, though, or in FEP-1b12. Is it documented anywhere else? Do other fediverse groups projects like Mbin/Kbin do it too?

No the documentation only covers the JSON format of messages, not how exactly they are processed (as that can change often and would require too much work).

I cant say how it works for other platforms, going to tag the respective developers instead:
- Mbin: @melroy89 
- Piefed: @rimu 
- NodeBB: @julianlam

**Comment by julianlam** [comment]
NodeBB uses "first line" title generation as well. It sucks, but it's the best we can do considering microblogs do not send titles.

~Here is our specific logic:~
~https://github.com/NodeBB/NodeBB/blob/master/src/activitypub/helpers.js#L343~

I was wrong, we don't use this anymore. We just prettify the html, call a tokenizer to split it into sentences, and grab the first sentence.

**Comment by julianlam** [comment]
Back to the original topic, while NodeBB federates well with BridgyFed, opt in is by user DM (private note). If BridgyFed were to introduce an instance-wide opt-in mechanism, I would be happy to adopt it.

**Comment by snarfed** [comment]
@julianlam glad to hear it's working well with NodeBB!

And we already have an instance-wide opt-in mechanism, added it way back last year, https://fed.brid.gy/docs#enable-api (https://github.com/snarfed/bridgy-fed/issues/1305#issuecomment-2874999193). There's generally no way for us to automatically discover new users on a given fediverse instance, so it takes a bit of work on admins' part, but hopefully not too bad. Feel free to point any admins our way who want to try!

**Comment by snarfed** [comment]
@Nutomic aha, good point on `\n` vs `<br>`, thanks for pointing that out. We originally depended on the fact that most fediverse servers render posts with `white-space: pre-wrap` CSS, https://github.com/snarfed/bridgy-fed/issues/990#issuecomment-2121071809, but that obviously doesn't work for non-rendering use cases like this. We can definitely tackle better whitespace handling as part of this!

Btw, if this first-line logic is a load bearing part of fediverse groups interop, should we consider adding it to FEP-1b12?

**Comment by julianlam** [comment]
It's not part of groups federation, threadiverse software sends titles, so no title generation needed. 🙂

**Comment by Nutomic** [comment]
> And we already have an instance-wide opt-in mechanism, added it way back last year, https://fed.brid.gy/docs#enable-api (https://github.com/snarfed/bridgy-fed/issues/1305#issuecomment-2874999193). There's generally no way for us to automatically discover new users on a given fediverse instance, so it takes a bit of work on admins' part, but hopefully not too bad. Feel free to point any admins our way who want to try!

I found that you can simply enable bridging for an instance via [this page](https://fed.brid.gy/web-site) (which does the same as `HTTP POST to https://fed.brid.gy/web-site` mentioned in the docs above). It seems to [work](https://bsky.app/profile/voyager.lemmy.ml.web.brid.gy), even though there is no validation that the instance was actually added by an admin.

> Btw, if this first-line logic is a load bearing part of fediverse groups interop, should we consider adding it to FEP-1b12?

1b12 is already final so it cannot be changed. That said if you write a new FEP with this information it would definitely be helpful!

**Comment by snarfed** [comment]
Ah, sorry for the confusion. Bridgy Fed currently bridges three different networks: Bluesky, the fediverse, and the web. The `/web-site` form/endpoint is for bridging a web site, via webmention and/or RSS/Atom, not for bridging a fediverse instance. https://fed.brid.gy/docs#web-get-started

We usually detect fediverse instances and block adding them this way, but that detection currently depends on both webfinger and `/.well-known/host-meta`, #1685, and it looks like Lemmy (or at least voyager.lemmy.ml) doesn't currently serve host-meta, so we didn't detect it.

**Comment by snarfed** [comment]
For now, I've disabled voyager.lemmy.ml. I can re-enable it if you really want to bridge it just from its RSS feed, but that's probably not the right direction.

**Comment by snarfed** [comment]
Amusingly, lemmy.ml also got bridged this way, a while ago: https://bsky.app/profile/lemmy.ml.web.brid.gy . It has followers, so I haven't disabled it, but I can if you want.

**Comment by snarfed** [comment]
Btw @Nutomic it looks like Lemmy is currently blocking requests from Bluesky's AppView/CDN, which is why https://bsky.app/profile/lemmy.ml.web.brid.gy 's profile picture is blank. Background in https://fed.brid.gy/docs#bluesky-image-cdn . Example CDN URL that says _Source image is unreachable_: https://cdn.bsky.app/img/avatar/plain/did:plc:q2k4ilmlzzrnoog5dccpqwor/bafkreibjocl6d6trnndsbt2g5bb3iwwqp5yqbrcs7vt776ecbhsmyaz4ry@jpeg

**Comment by Nutomic** [comment]
We block requests with empty user-agent and some bot user-agents from crawlers. Its possible that Bluesky is caught in those. Do you know what user-agent it uses? Anyway its better to use voyager.lemmy.ml for testing purposes (it doesnt have these user-agent blocks).

**Comment by snarfed** [comment]
Historically their CDN may have omitted User-Agent entirely. They tried to fix that, but maybe only partially, maybe not on redirects? I'm not sure. Details in https://github.com/bluesky-social/atproto/discussions/3451

**Comment by FurblandChannel** [comment]
It might be worthwhile to at least *begin* with having lemmy API auth, so you can log in with a lemmy account (or multiple) as the admin of a community and use those perms to prove you have authority to set up the bridge?

**Comment by snarfed** [comment]
@FurblandChannel hmm! Auto-enabling the bridge for entire instances (https://fed.brid.gy/docs#enable-api) is hopefully orthogonal to FEP-1b12 groups support, right?

**Comment by FurblandChannel** [comment]
Maybe? I was more talking about individual communities. You log in with a Lemmy account that moderates a community, and those permissions are used as authorization to enable the bridge for that community

**Comment by snarfed** [comment]
Ah, ok. Groups (communities) aren't really analogous to instances, true, which also means that I don't think we can bridge entire groups, at least not if it means automatically enabling the bridge for everyone who posts to the group. I think that opt in still needs to be up to individual users and/or instance admins.

**Comment by FurblandChannel** [comment]
It wouldn’t necessarily opt all the users in, just the group. bsky users could then post to the group across the bridge, and posts from bridged fedi users would be visible as being reblogged by the group on bsky

**Comment by FurblandChannel** [comment]
…so that way you could opt in groups that you own, and then the users in those groups would have to opt in individually to appear in the corresponding group on bsky

> On Mar 2, 2026, at 11:44 AM, Furbland wrote:
> 
> ﻿
> It wouldn’t necessarily opt all the users in, just the group. bsky users could then post to the group across the bridge, and posts from bridged fedi users would be visible as being reblogged by the group on bsky 
>

**Comment by snarfed** [comment]
Ah! If we'd only expect already-bridged users' posts to be reposted by the group, then sure. That's just bridging the group in general; I think this issue covers that.

**Comment by FurblandChannel** [comment]
yeah, I was just suggesting a way to prove that you own the group in order to bridge it (oauth-ing with an account that moderates said group)

**Comment by Die4Ever** [comment]
any thoughts about this? https://www.theverge.com/tech/948215/bluesky-communities-at-protocol-atmosphere-reddit

**Comment by snarfed** [comment]
Hah, yes! I think that's orthogonal. So far, we've always drawn a pretty bright line at only bridging fully public data. Given that, I'm only really thinking about public groups and posts here. Bluesky communities, on the other hand, are more oriented toward private or at least non-public data, at least as far as we know.

Still early though. We'll see!

**Comment by BrettM86** [comment]
> More broadly though, microblogging is obviously pretty different from groups/forums, I expect there's a limit on how well we'll be able to bridge from/to Bluesky specifically. We haven't done a deep dive into ATProto group/forum apps, but those might be more appropriate for bridging, [#1178](https://github.com/snarfed/bridgy-fed/issues/1178) notwithstanding. Feel free to look and let us know what you find!


For anyone who's curious and looking for a more 1:1 mirror of groups/forums I have been working on an AP threadiverse↔atproto bridge over at Tidepool (https://github.com/BrettM86/tidepool) (GH Mirror). Still in the early stages, but follows FEP-1b12 communities and uses social.coves.* records (a forum-shaped atproto app I'm also building). It gets around the microblogging lexicon limitations since it's intended as a forum platform. 

It's opt-in by default, but it respects the typical Lemmy defederation (tdpl.io for the read mirror, coves.social for the native content -- needs both for a full atproto defederate), but we also respect #nobridge / #nobot for users.

Right now it's live mirroring a few pre-selected AP communities (posts, comments, votes); outbound we have votes and comments federated properly, posts are implemented but I haven't tested them against a real instance yet. Native Coves communities federating is probably the next step, but a ways out.

**Comment by rimu** [comment]
Probably worth taking the temperature of the room at lemmy.world/c/fediverse to see how people feel about being bridged by default and needing to opt-out to not be bridged.

Generally the people on Lemmy are less uptight about this kind of thing ("all activitypub data is public" is a common saying among them) than on Mastodon so I think it'd go ok. But good to check first.

**Comment by julianlam** [comment]
@BrettM86 happy to provide a NodeBB group for integration testing. Use @general@activitypub.space
